sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDbStateStoreProvider.scala

```diff
@@ -0,0 +1,652 @@
+    } catch {
+      case NonFatal(e) =>
+        logWarning(s"Error performing snapshot and cleaning up $this")
+    }
+  }
+
+  private def doSnapshot(files: Seq[WALUtils.StoreFile]): Unit = {
+    if (files.nonEmpty) {
+      val lastVersion = files.last.version
+      val deltaFilesForLastVersion =
+        filesForVersion(files, lastVersion).filter(_.isSnapshot == false)
+      if (deltaFilesForLastVersion.size > storeConf.minDeltasForSnapshot) {
+        val dbPath = getBackupPath(lastVersion)
+        val snapShotFileName = getTempPath(lastVersion).concat(".snapshot")
+        val f = new File(snapShotFileName)
+        f.delete() // delete any existing tarball
+        try {
+          val (_, t1) = Utils.timeTakenMs {
+            FileUtility.createTarFile(dbPath, snapShotFileName)
+            val targetFile = snapshotFile(baseDir, lastVersion)
+            uploadFile(fm, new Path(snapShotFileName), targetFile, sparkConf)
+          }
+          logInfo(
+            s"Creating snapshot file for" +
+              s" ${stateStoreId_.partitionId} took $t1 ms.")
+        } catch {
+          case e: Exception =>
+            logError(s"Exception while creating snapshot $e} ")
+            throw e
+        } finally {
+          f.delete() // delete the tarball
+        }
+      }
+    }
+  }
+
+  /*
+   * Clean up old snapshots and delta files that are not needed any more. It ensures that last
+   * few versions of the store can be recovered from the files, so re-executed RDD operations
+   * can re-apply updates on the past versions of the store.
+   */
+  private[state] def cleanup(files: Seq[WALUtils.StoreFile]): Unit = {
+    try {
+      if (files.nonEmpty) {
+        val earliestVersionToRetain = files.last.version - storeConf.minVersionsToRetain
+        if (earliestVersionToRetain > 0) {
+          val earliestFileToRetain = filesForVersion(files, earliestVersionToRetain).head
+          val filesToDelete = files.filter(_.version < earliestFileToRetain.version)
+          val (_, e2) = Utils.timeTakenMs {
+            filesToDelete.foreach { f =>
+              fm.delete(f.path)
+              val file = new File(rocksDbPath, f.version.toString.toUpperCase(Locale.ROOT))
+              if (file.exists()) {
+                file.delete()
+              }
+            }
+          }
+          logDebug(s"deleting files took $e2 ms.")
+          logInfo(
+            s"Deleted files older than ${earliestFileToRetain.version} for $this: " +
+              filesToDelete.mkString(", "))
+        }
+      }
+    } catch {
+      case NonFatal(e) =>
+        logWarning(s"Error cleaning up files for $this", e)
+    }
+  }
+
+  private def cleanRocksDBBackupInstances(files: Seq[WALUtils.StoreFile]): Unit = {
+    try {
+      if (files.nonEmpty) {
+        val earliestVersionToRetain = files.last.version - storeConf.minVersionsToRetain
+        if (earliestVersionToRetain > 0) {
+          for (v <- (earliestVersionToRetain - 1) to 1) {
```

Comment:
Scala is not that enough smart to automatically create decreasing range. 
`(1 to (earliestVersionToRetain -1)).reverse` or `((earliestVersionToRetain -1) to 1 by -1)` will do the trick.

Reply:
Btw, what if version is going to be pretty huge like 1000 or so? I know this is simpler, but you may still need to list the directory and find things to delete.